# 프로젝트(1) load_digits: 손글씨를 분류해봅시다.

## 필요한 모듈 import하기

In [1]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

## 데이터 준비

In [2]:
digits = load_digits()
digits.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

## 데이터 이해하기

In [3]:
#feature data 지정하기
digits_data = digits.data
digits_data.shape

(1797, 64)

In [4]:
#label data 지정하기
digits_label = digits.target
print(digits_label.shape)

(1797,)


In [5]:
# target names 출력해보기
print(digits.target_names)

[0 1 2 3 4 5 6 7 8 9]


In [6]:
# 데이터 describe헤보기
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

## train, test 데이터 분리

In [7]:
X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                    digits_label,
                                                    test_size=0.2,
                                                    random_state = 15)

## 다양한 모델로 학습시켜보기

In [8]:
# Decision Tree
decision_tree = DecisionTreeClassifier(random_state = 15)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.8444444444444444


              precision    recall  f1-score   support

           0       1.00      0.90      0.95        31
           1       0.84      0.82      0.83        38
           2       0.73      0.87      0.80        38
           3       0.71      0.74      0.73        27
           4       0.94      0.80      0.87        41
           5       0.87      0.94      0.90        35
           6       0.87      0.89      0.88        38
           7       0.91      0.94      0.93        34
           8       0.79      0.77      0.78        35
           9       0.80      0.77      0.79        43

    accuracy                           0.84       360
   macro avg       0.85      0.85      0.84       360
weighted avg       0.85      0.84      0.85       360



In [10]:
#Random Forest
random_forest = RandomForestClassifier(random_state = 30)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9722222222222222


              precision    recall  f1-score   support

           0       0.97      0.94      0.95        31
           1       0.95      1.00      0.97        38
           2       1.00      1.00      1.00        38
           3       0.93      0.96      0.95        27
           4       0.95      0.98      0.96        41
           5       1.00      1.00      1.00        35
           6       1.00      0.95      0.97        38
           7       0.97      1.00      0.99        34
           8       0.97      0.91      0.94        35
           9       0.98      0.98      0.98        43

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



In [11]:
#svm
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9833333333333333


              precision    recall  f1-score   support

           0       1.00      0.97      0.98        31
           1       0.95      1.00      0.97        38
           2       1.00      1.00      1.00        38
           3       0.96      0.96      0.96        27
           4       0.98      0.98      0.98        41
           5       1.00      1.00      1.00        35
           6       1.00      1.00      1.00        38
           7       1.00      1.00      1.00        34
           8       0.97      0.94      0.96        35
           9       0.98      0.98      0.98        43

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [12]:
#SGD Classifier 
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9388888888888889


              precision    recall  f1-score   support

           0       1.00      0.97      0.98        31
           1       0.95      0.92      0.93        38
           2       1.00      0.95      0.97        38
           3       0.93      0.93      0.93        27
           4       0.95      0.98      0.96        41
           5       0.95      1.00      0.97        35
           6       1.00      0.84      0.91        38
           7       0.97      1.00      0.99        34
           8       0.72      0.97      0.83        35
           9       1.00      0.86      0.92        43

    accuracy                           0.94       360
   macro avg       0.95      0.94      0.94       360
weighted avg       0.95      0.94      0.94       360



In [13]:
#Logistic Regression
logistic_model = LogisticRegression(C=10, tol=0.01, solver='lbfgs',max_iter=1000)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9638888888888889


              precision    recall  f1-score   support

           0       1.00      0.97      0.98        31
           1       0.95      0.97      0.96        38
           2       0.97      0.97      0.97        38
           3       0.96      0.93      0.94        27
           4       0.95      0.98      0.96        41
           5       0.94      0.97      0.96        35
           6       1.00      0.97      0.99        38
           7       1.00      1.00      1.00        34
           8       0.89      0.91      0.90        35
           9       0.98      0.95      0.96        43

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



## 모델을 평가해보기

1. 이번 분석에 쓰인 데이터의 경우, 데이터의 타겟의 수와 데이터의 클래스의 개수가 같기 때문에 데이터 분포의 심각한 불균형은 거의 없다. 그러므로 이번 학습에서는 전체 데이터 중 올바르게 판단한 데이터 개수의 비율인 accuracy가 학습의 결과를 잘 나타낸다고 할 수 있다. 그러므로, 
    - Decision Tree:0.84
    - Random Forest:0.972
    - SVM:0.983
    - SGD classifier:0.938
    - Logistic Regression:0.963
    
   위의 결과에서 SVM, Logistic regression,Random Forest순으로 학습 결과가 좋게 나타난다. 
2. Decision Tree제외  0.9이상의 높은 정확도를 보였다.  


# 프로젝트(2) load_wine: 와인을 분류해봅시다.

## 필요한 모듈 import하기

In [14]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

 ## 데이터 준비

In [41]:
wine = load_wine()
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

## 데이터 이해하기

In [42]:
#feature data 지정하기
wine_data = wine.data
wine_data.shape

(178, 13)

In [43]:
#label data 지정하기
wine_label= wine.target
wine_label.shape

(178,)

In [44]:
#target names 출력해보기
print(wine.target_names)

['class_0' 'class_1' 'class_2']


In [19]:
#data describe 해보기
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

 ## train, test 데이터 분리

In [57]:
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size=0.2,
                                                    random_state = 12
                                                   )

## 다양한 모델로 학습 시켜보기

In [58]:
# Decision Tree
decision_tree = DecisionTreeClassifier(random_state = 15)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9166666666666666


              precision    recall  f1-score   support

           0       0.82      1.00      0.90        14
           1       1.00      0.91      0.95        11
           2       1.00      0.82      0.90        11

    accuracy                           0.92        36
   macro avg       0.94      0.91      0.92        36
weighted avg       0.93      0.92      0.92        36



In [59]:
#Random Forest
random_forest = RandomForestClassifier(random_state = 30)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9722222222222222


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      0.91      0.95        11
           2       0.92      1.00      0.96        11

    accuracy                           0.97        36
   macro avg       0.97      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



In [60]:
#svm
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.6388888888888888


              precision    recall  f1-score   support

           0       1.00      0.86      0.92        14
           1       0.50      1.00      0.67        11
           2       0.00      0.00      0.00        11

    accuracy                           0.64        36
   macro avg       0.50      0.62      0.53        36
weighted avg       0.54      0.64      0.56        36



In [61]:
#SGD Classifier 
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.75


              precision    recall  f1-score   support

           0       1.00      0.86      0.92        14
           1       0.75      0.55      0.63        11
           2       0.56      0.82      0.67        11

    accuracy                           0.75        36
   macro avg       0.77      0.74      0.74        36
weighted avg       0.79      0.75      0.76        36



In [62]:
#Logistic Regression
logistic_model = LogisticRegression(C=10, tol=0.01, solver='lbfgs',max_iter=10000)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9722222222222222


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      0.91      0.95        11
           2       0.92      1.00      0.96        11

    accuracy                           0.97        36
   macro avg       0.97      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



# 모델을 평가해보기

1. 위 다섯가지 모델의 aaccuracy를 비교해보면,
    
    - Decision Tree:0.91
    - Random Forest:0.97
    - SVM:0.63
    - SGD classifier:0.75
    - Logistic Regression:0.97
    
    Logistic Regression과 Random Forest가 0.97로 굉장히 높은 accuracy를 가진 다는 것을 알 수 있다. 
    
2. 하지만, 두가지 방법 모두 train_test_split함수의 random_state를 바꿔보면 acuracy가 1.0에 수렴하는 결과를 보였다. 이는 좋은 모델이라 단언할 수 없고, overfitting의 가능성이 높다. 
    이는 곧,
    training과 test set가 겹쳤다거나 하는 가설을 세워볼 수 있다. 
    
3. 그러므로, 0.91의 높은 정확성을 가지면서 random state에 영향을 받지 않는 Decision Tree 모델을 사용하는 것이 더욱 robust한 결과를 얻을 수 있을 것이다.


# 프로젝트(3) load_breast_cancer:유방암 여부를 진단해봅시다.

## 필요한 모듈 import

In [29]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

## 데이터 준비

In [30]:
breast_cancer = load_breast_cancer()
breast_cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

## 데이터 이해하기

In [31]:
#feature data 지정하기
cancer_data = breast_cancer.data
print(cancer_data.shape)

(569, 30)


In [32]:
#label data 지정하기
cancer_label = breast_cancer.target
print(cancer_label.shape)

(569,)


In [33]:
# target names 출력
print(breast_cancer.target_names)

['malignant' 'benign']


In [34]:
# 데이터 Describe해보기
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

## train, test 데이터 분리

In [35]:
X_train, X_test, y_train, y_test = train_test_split(cancer_data,
                                                    cancer_label,
                                                    test_size=0.2,
                                                    random_state = 12)

## 다양한 모델로 학습시켜보기

In [36]:
# Decision Tree
decision_tree = DecisionTreeClassifier(random_state = 15)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9473684210526315


              precision    recall  f1-score   support

           0       0.96      0.92      0.94        48
           1       0.94      0.97      0.96        66

    accuracy                           0.95       114
   macro avg       0.95      0.94      0.95       114
weighted avg       0.95      0.95      0.95       114



In [37]:
#Random Forest
random_forest = RandomForestClassifier(random_state = 30)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9298245614035088


              precision    recall  f1-score   support

           0       0.98      0.85      0.91        48
           1       0.90      0.98      0.94        66

    accuracy                           0.93       114
   macro avg       0.94      0.92      0.93       114
weighted avg       0.93      0.93      0.93       114



In [38]:
#svm
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.8859649122807017


              precision    recall  f1-score   support

           0       1.00      0.73      0.84        48
           1       0.84      1.00      0.91        66

    accuracy                           0.89       114
   macro avg       0.92      0.86      0.88       114
weighted avg       0.90      0.89      0.88       114



In [39]:
#SGD Classifier 
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.8947368421052632


              precision    recall  f1-score   support

           0       0.93      0.81      0.87        48
           1       0.88      0.95      0.91        66

    accuracy                           0.89       114
   macro avg       0.90      0.88      0.89       114
weighted avg       0.90      0.89      0.89       114



In [40]:
#Logistic Regression
logistic_model = LogisticRegression(C=10, tol=0.01, solver='lbfgs',max_iter=10000)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

#Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("ACCURACY : ",accuracy)
print("\n")

#confusion matrix scores
print(classification_report(y_test, y_pred))

ACCURACY :  0.9385964912280702


              precision    recall  f1-score   support

           0       0.98      0.88      0.92        48
           1       0.92      0.98      0.95        66

    accuracy                           0.94       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.94      0.94      0.94       114



# 모델을 평가해보기

1. 우선 다섯가지 모델의 Accuracy를 살펴본다면 

    - Decision Tree:0.94
    - Random Forest:0.92
    - SVM:0.88
    - SGD classifier:0.89
    - Logistic Regression:0.93
   
      의 결과로 Decision Tree, Logistic Regression, Random Forest가 0.9이상의 정확도를 낸다. 
      
2. 유방암을 판별한다는 특수성을 고려했을 떄, 양성인데 음성이라고 판별한 수가 낮아야 하는 RECALL점수를 유심히 살펴야 한다. 위 모델들 중에서는 Decision Tree의 RECALL 점수가 높으므로 Decision Tree 모델을 사용하는 것이 좋을 것이다. 